libbeat/common/kubernetes/k8skeystore/kubernetes_keystore.go

```diff
@@ -76,11 +79,16 @@ func (kr *KubernetesKeystoresRegistry) GetKeystore(event bus.Event) keystore.Key
 	}
 	if namespace != "" {
 		// either retrieve already stored keystore or create a new one for the namespace
+		kr.keystoreMapLock.RLock()
 		if storedKeystore, ok := kr.kubernetesKeystores[namespace]; ok {
+			kr.keystoreMapLock.RUnlock()
 			return storedKeystore
 		}
+		kr.keystoreMapLock.RUnlock()
 		k8sKeystore, _ := Factoryk8s(namespace, kr.client, kr.logger)
-		kr.kubernetesKeystores["namespace"] = k8sKeystore
+		kr.keystoreMapLock.Lock()
```

Comment:
This operation is not atomic, so multiple keystore instances could be created if multiple goroutines compete for this resource.
If this is a problem, we should also check if the key exists in the write lock, before instantiating a new keystore and storing it in the map.

Reply:
Ah makes sense!
I think it should be better to apply the lock on the whole read/write block (https://github.com/elastic/beats/pull/21880/commits/21d61b17c60445d1b29e1a6da47450be99663f42). Let me know what you think.

Reply:
Take into account that a write lock produces a higher contention that a read lock. If this is in a critical path, it could affect performance. If that is the case, I would be ok with duplicating the code to check if the keystore exist.

Reply:
Also if performance can be an issue here, consider using a `sync.Map`, that is optimized for entries that are written once but read many times: https://golang.org/pkg/sync/#Map
For this case we would probably need to use `LoadOrStore`, storing an object that lazily creates the keystore.

Reply:
👍